Jsonparser step5

junibibi opened the pull request:
7-5. JSON 문자열 생성 진행했습니다.

JurdyPark committed:
Serialize 구조체생성 하고, jsonValue를 직렬화하는 메소드 껍데기작성함

JurdyPark committed:
indentCount에따라 공백이 추가되는 코드 작성

JurdyPark committed:
serializeArray메소드 수정
-하드코딩한 공백자리에 공백을추가해주는 메소드를 호출함
-공백이 추가되는 위치에 공백카운트를 변화시켜주는 코드 추가함
-직렬화될 값들이 serializeValue메소드를 통해 들어도록 수정함

JurdyPark committed:
serializeValue메소드 수정함
-case를 array와 아닌경우 두가지로 분리함
(array가 아닌 경우는 일반적으로 print하면 되기 때문에..)
-매개변수명 수정함

JurdyPark committed:
serializeObject 메소드 생성함

JurdyPark committed:
구조체/파일이름 수정함

JurdyPark committed:
Serializer 인스턴스 생성함
do코드 수정함
 - do문밖에 jsonData를 상수로 선언하고, catch문에 return을 추가하여 예외처리 끝나고 jsonData를 사용할수 있게함
직렬화된 jsonData출력하는 코드 추가함

JurdyPark committed:
serializeValue에서 serializeObject호출하는 경우 추가함

JurdyPark committed:
serializeArray에서 들여쓰기 하지 않도록 수정함.
공백정리함

godrm commented on JSONParser/JSONParser/Serializer.swift:
```diff
@@ -0,0 +1,77 @@
+        case let object as [String: JsonType]:
+            return serializeObject(object)
+        case let array as [JsonType]:
+            return serializeArray(array)
```
이렇게 생각해보세요.
함수에 매개변수를 전달해서 그 값으로만 리턴값을 만들어내는 경우
`serializeArray(_ array: [JsonType]) -> String` 는  `array.serialize() -> String` 형태가 가능합니다.
마찬가지로 object도 같은 형식이 가능하구요.
그렇게 되면 타입 비교없이 jsonData.serialize()로 호출할 수 있습니다. 
동일한 형식으로 만들어진 함수를 다형성을 통해서 호출할 수 있는거죠.
이렇게 개선해보세요.

JurdyPark committed:
JsonType이 Serializable,Countable 프로토콜을 채택하도록 함.
Double,String,Bool타입이 JsonType프로토콜을 채택하도록 수정함.
string의 serialized는 반환할때 앞뒤로 "가 추가되도록 수정함.

JurdyPark committed:
Array가 채택한 프로토콜별로 extension 해줌

JurdyPark committed:
오브젝트가 채택한 프로토콜별로 extension 해줌

JurdyPark committed:
프로토콜을 구현할때 필요한 변수들을 상위로 위치변경함.
Serializable프로토콜 선언함.
Serializable기본구현함.
Array와 오브젝트가 Serializable프로토콜을 채택하고 extention해줌
Serializer구조체 위치를 아래로 변경함.

junibibi commented:
피드백반영했습니다.

godrm commented:
수고하셨습니다. 
마지막 수정사항 중에서 서로 다른 프로토콜에 대해 상속 관계를 가질 때는 두 프로토콜이 매우 밀첩한 관계여야 합니다.
JsonType 이지만 Countable을 채택하지 않아야 하는 경우가 있다면 상속을 하면 안됩니다. 이 부분은 꼭 기억하세요.
